core/src/main/java/org/apache/spark/shuffle/sort/ShuffleInMemorySorter.java

```diff
@@ -32,24 +37,39 @@ public int compare(PackedRecordPointer left, PackedRecordPointer right) {
   }
   private static final SortComparator SORT_COMPARATOR = new SortComparator();
 
+  private final MemoryConsumer consumer;
+  private final TaskMemoryManager memoryManager;
+
   /**
    * An array of record pointers and partition ids that have been encoded by
    * {@link PackedRecordPointer}. The sort operates on this array instead of directly manipulating
    * records.
    */
-  private long[] array;
+  private LongArray array;
 
   /**
    * The position in the pointer array where new records can be inserted.
    */
   private int pos = 0;
 
-  public ShuffleInMemorySorter(int initialSize) {
+  public ShuffleInMemorySorter(
+      MemoryConsumer consumer,
+      TaskMemoryManager memoryManager,
+      int initialSize) {
+    this.consumer = consumer;
+    this.memoryManager = memoryManager;
     assert (initialSize > 0);
-    this.array = new long[initialSize];
+    this.array = new LongArray(memoryManager.allocatePage(initialSize * 8L, consumer));
```

Comment:
If this `allocatePage` call were to fail, I think you'd get an NPE here, since `allocatePage` would return null.